utilities/postgres.go

```diff
@@ -0,0 +1,67 @@
+package utilities
+
+import (
+	"errors"
+	"strconv"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgerrcode"
+)
+
+// PostgresError is a custom error struct for marshalling Postgres errors to JSON.
+type PostgresError struct {
+	Code           string `json:"code"`
+	HttpStatusCode int    `json:"-"`
+	Message        string `json:"message"`
+	Hint           string `json:"hint,omitempty"`
+	Detail         string `json:"detail,omitempty"`
+}
+
+// NewPostgresError returns a new PostgresError if the error was from a publicly
+// accessible Postgres error.
+func NewPostgresError(err error) *PostgresError {
+	var pgErr *pgconn.PgError
+	if errors.As(err, &pgErr) && isPubliclyAccessiblePostgresError(pgErr.Code) {
+		return &PostgresError{
+			Code:           pgErr.Code,
+			HttpStatusCode: getHttpStatusCodeFromPostgresErrorCode(pgErr.Code),
+			Message:        pgErr.Message,
+			Detail:         pgErr.Detail,
+			Hint:           pgErr.Hint,
+		}
+	}
+
+	return nil
+}
+
+// isPubliclyAccessiblePostgresError checks if the Postgres error should be
+// made accessible.
+func isPubliclyAccessiblePostgresError(code string) bool {
+	if len(code) != 5 {
+		return false
+	}
+
+	if code == pgerrcode.RaiseException {
+		return true
+	}
+
+	return getHttpStatusCodeFromPostgresErrorCode(code) != 0
+}
+
+// getHttpStatusCodeFromPostgresErrorCode maps a Postgres error code to a HTTP
+// status code.
+func getHttpStatusCodeFromPostgresErrorCode(code string) int {
+	if code == pgerrcode.RaiseException {
+		return 500
+	}
+
+	// Use custom HTTP status code if Postgres error was triggered with `PTXXX`
+	// code. This is consistent with PostgREST's behaviour as well.
+	if code[0:2] == "PT" {
+		if httpStatusCode, err := strconv.ParseInt(code[2:], 10, 0); err == nil {
+			return int(httpStatusCode)
+		}
+	}
+
```

Comment:
can we also check for integrity constraint violations here? (e.g. unique / foreign key violation)

some context: @silentworks and I tested this out and realised that while the error from a "raise exception" in a before / after trigger gets bubbled up, if the trigger returns an error that is not from a "raise exception", then gotrue defaults to return an internal server error.

Reply:
I only permitted the `raise exception` and `raise sqlstate` errors to bubble up because that was what was suggested in #271. We could easily permit all errors to bubble up, which might be helpful especially if there are other types of errors. However, because the error details will be exposed for all HTTP errors returned with an internal error, we probably need to be more specific about which ones to expose.

I can update the PR to add the integrity constraint violations to the acceptable types of Postgres errors. (Just wondering, why those specifically?) Any opinions on other types of errors to expose, and on what status code to return? (I guess 500 would still be appear to be the most appropriate one.)

I can probably add some tests using triggers sometime this week (similar to https://github.com/supabase/gotrue/pull/405/commits/58de2b721092505224d8b0b68fe14afc8dc1c03b).

Reply:
> I can update the PR to add the integrity constraint violations to the acceptable types of Postgres errors. (Just wondering, why those specifically?)

Yeah the main motivation for this is to help catch errors that don't get caught in a `raise exception` clause. If the trigger written fails to consider a unique / foreign key constraint and fails, gotrue will return a 500 internal server error which isn't that useful for a developer when it comes to debugging. 

Now that we've exposed the auth logs on the logs explorer in the dashboard, they should be able to see the actual error behind the 500 but i still think it's useful to return it explicitly since there's an appropriate http error code mapped to the error returned by postgres. 

In terms of other types of errors to expose, i did a quick look at the list and felt that the integrity constraint violations were the most obvious to include, but feel free to make a suggestion if you feel that I'm missing out anything :)

Reply:
I was thinking maybe errors for invalid data types would be helpful to expose, but I would need to take a deeper look into which errors those would be in the list.